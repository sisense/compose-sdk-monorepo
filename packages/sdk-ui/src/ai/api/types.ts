import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
