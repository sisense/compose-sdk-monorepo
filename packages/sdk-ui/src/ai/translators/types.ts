import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '@/ai';
import { ChartType, TableType } from '@/types';
import { WidgetModel } from '@/models';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * All Chart Types including Table
 * Note: Table will be re-united with ChartType in a future MR
 *
 * @internal
 */
export type AllChartType = ChartType | TableType;

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart or table widget in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * Naming TBD
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps' | 'getTableWidgetProps'> & {
  chartType: AllChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl' | 'tableWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
