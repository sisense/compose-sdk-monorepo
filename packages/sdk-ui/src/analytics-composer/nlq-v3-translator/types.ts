import { Filter, FilterRelations, JSONArray, JSONValue, Measure } from '@sisense/sdk-data';

import { DataSchemaContext, NlqTranslationInput } from '../types';

export const ATTRIBUTE_PREFIX = 'DM.';

// Argument types for function validation
export type ArgType =
  // Basic types
  | 'string' // "Total Revenue", "DESC", etc.
  | 'string[]' // ["2012-01-01T00:00:00"], member arrays
  | 'number' // 5, 100, ranking values
  | 'number[]' // [10, 20] for numeric ranges
  | 'boolean' // true/false flags

  // SDK Object types (from function calls)
  | 'Attribute' // "DM.Commerce.Revenue" → Attribute
  | 'Attribute[]' // ["DM.Commerce.AgeRange"] → Attribute[]
  | 'Measure' // Function call → Measure (general)
  | 'BaseMeasure' // Function call → BaseMeasure (specific aggregation)
  | 'Measure[]' // Array of function calls → Measure[]
  | 'Filter' // Function call → Filter
  | 'Filter[]' // Array of function calls → Filter[]
  | 'FilterRelationsNode' // Filter | FilterRelations | Filter[]
  | 'DateDimension' // Date-specific attributes
  | 'LevelAttribute' // Date level attributes

  // Union types
  | 'Measure | number' // Can be either measure function or number
  | 'Date | string' // Date objects or ISO date strings
  | 'string | number' // String or numeric values

  // Configuration objects
  | 'BaseFilterConfig' // Optional filter config
  | 'MembersFilterConfig' // Members-specific filter config
  | 'CustomFormulaContext' // Custom formula context object

  // Catch-all for complex objects
  | 'any'; // Skip validation for complex types

// Argument schema for function validation
export interface ArgSchema {
  type: ArgType;
  required: boolean;
}

// Type to represent a function argument
export type Arg = string | number | FunctionCall | string[];
export type ProcessedArg = unknown;
export type QueryElement = Filter | FilterRelations | Measure;
export type FactoryFunction = (...args: unknown[]) => QueryElement;

// Type to represent the structure parsed from the JSON
export type FunctionCall = {
  function: string;
  args: Arg[];
};

/**
 * Context object passed to all custom processors containing runtime information
 * needed for processing.
 */
export interface FunctionContext extends DataSchemaContext {
  /** Path prefix for error messages (e.g., 'args[0].function') */
  pathPrefix: string;
}

export interface ArgContext extends FunctionContext {
  argSchema: ArgSchema;
}

// Translation input types
export type DimensionsInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type MeasuresInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type FiltersInput = NlqTranslationInput<JSONArray, DataSchemaContext>;
export type HighlightsInput = NlqTranslationInput<JSONArray, DataSchemaContext>;

// Function call input types
export type MeasuresFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;
export type FiltersFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;
export type HighlightsFunctionCallInput = NlqTranslationInput<FunctionCall[], DataSchemaContext>;

// Core processing input types
export type NodeInput = NlqTranslationInput<FunctionCall, FunctionContext>;
export type ArgInput = NlqTranslationInput<Arg, ArgContext>;

/**
 * Custom processor function signature.
 */
export type CustomFunctionProcessor = (
  processedArgs: ProcessedArg[],
  context: FunctionContext,
) => void;

export function isFunctionCall(value: JSONValue): value is FunctionCall {
  return typeof value === 'object' && value !== null && 'function' in value && 'args' in value;
}

export function isFunctionCallArray(value: JSONArray): value is FunctionCall[] {
  return value.every(isFunctionCall);
}

// Type guard to check if the argument is a Filter
export function isFilterElement(arg: QueryElement): arg is Filter {
  return 'attribute' in arg && 'config' in arg;
}

// Type guard to check if the argument is a FilterRelations
export function isFilterRelationsElement(arg: QueryElement): arg is FilterRelations {
  return 'left' in arg && 'right' in arg && 'operator' in arg;
}

export function isMeasureElement(arg: QueryElement): arg is Measure {
  return !isFilterElement(arg) && !isFilterRelationsElement(arg);
}

export function isStringArray(value: JSONArray): value is string[] {
  return value.every((item) => typeof item === 'string');
}
