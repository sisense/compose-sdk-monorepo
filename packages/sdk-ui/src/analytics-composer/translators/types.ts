import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '../../ai';
import { ChartType, WidgetModel } from '../../';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart widget (including table) in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps'> & {
  chartType: ChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
