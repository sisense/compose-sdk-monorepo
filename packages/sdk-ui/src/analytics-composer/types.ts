import { WidgetProps, ChartWidgetProps, PivotTableWidgetProps } from '@/props';
import { ChartDataOptions, ChartStyleOptions } from '@/types';
import { ExecuteQueryParams, ExecutePivotQueryParams } from '@/query-execution';
import { JSONArray, MetadataItem } from '@sisense/sdk-data';
import { DynamicChartType } from '../chart-options-processor/translations/types.js';
import { AnyColumn } from '@/chart-data-options/types.js';
import { DashboardProps } from '@/dashboard/types.js';

type Stringify<T> = {
  [K in keyof T as `${K & string}String`]: string;
};

type ExtraWidgetCodeProps = {
  componentString: string;
  extraImportsString: string;
  idString?: string;
  widgetTypeString?: 'chart' | 'pivot' | 'text' | 'custom';
};

/** @internal */
export type KeysOfUnion<T> = T extends T ? keyof T : never;
/** @internal */
export type AllPossibleChartOptionKeys = KeysOfUnion<ChartDataOptions>;
/** @internal */
export type AxesMappingKey = Exclude<AllPossibleChartOptionKeys, 'seriesToColorMap'>;
/** @internal */
export type AxesMapping = Partial<Record<AxesMappingKey, Array<AnyColumn>>>;

/** @internal */
export interface ChartRecommendations {
  chartFamily: string;
  chartType: string;
  axesMapping: AxesMapping;
  styleOptions?: ChartStyleOptions;
}

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export interface ExpandedQueryModel {
  /** @internal */
  chartRecommendations: ChartRecommendations | {};
  /** @internal */
  jaql: {
    datasource: {
      id?: string;
      type?: 'elasticube' | 'live';
      title: string;
    };
    metadata: MetadataItem[];
  };
  /** @internal */
  queryTitle: string;
}

/**
 * Empty Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '', type: 'elasticube' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
  styleOptions: ChartStyleOptions;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * QueryTranslator is responsible for translating this query model to ExpandedWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey =
  | 'baseChartTmpl'
  | 'chartTmpl'
  | 'chartWidgetTmpl'
  | 'chartWidgetPropsTmpl'
  | 'widgetByIdTmpl'
  | 'executeQueryByWidgetIdTmpl'
  | 'executeQueryTmpl'
  | 'executePivotQueryTmpl'
  | 'pivotTableWidgetTmpl'
  | 'pivotTableWidgetPropsTmpl'
  | 'dashboardByIdTmpl'
  | 'dashboardTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};

/**
 * Code Placeholder Map
 *
 * @internal
 */
export type CodePlaceholderMap = Record<string, string>;

/**
 * Base Code Config
 *
 * @internal
 */
export type BaseCodeConfig = { uiFramework?: UiFramework };

/** TYPES FOR WIDGET CODE */

/**
 * Widget Code Config
 *
 * @internal
 */
export type WidgetCodeConfig = BaseCodeConfig & { includeChart?: boolean };

/**
 * Client-side Widget Code Params
 *
 * @internal
 */
export type ClientSideWidgetCodeParams = BaseCodeConfig & {
  widgetProps: WidgetProps;
};

/**
 * By ID Widget Code Params
 *
 * @internal
 */
export type ByIdWidgetCodeParams = WidgetCodeConfig & {
  dashboardOid: string;
  widgetOid: string;
  chartType: DynamicChartType;
};

/**
 * Widget Code Params
 *
 * @internal
 */
export type WidgetCodeParams = ClientSideWidgetCodeParams | ByIdWidgetCodeParams;

/**
 * Check if widget code params is for client-side code
 *
 * @internal
 */
export const isClientSideWidgetCodeParams = (
  params: WidgetCodeParams,
): params is ClientSideWidgetCodeParams => {
  return 'widgetProps' in params;
};

/**
 * Check if widget code params is for by ID code
 *
 * @internal
 */
export const isByIdWidgetCodeParams = (
  params: WidgetCodeParams,
): params is ByIdWidgetCodeParams => {
  return 'dashboardOid' in params && 'widgetOid' in params;
};

/**
 * Widget Props Config
 *
 * @internal
 */
export type WidgetPropsConfig = { useCustomizedStyleOptions?: boolean };

/**
 * Execute Query Code Params
 *
 * @internal
 */
export type ExecuteQueryCodeParams = BaseCodeConfig & {
  queryParams: ExecuteQueryParams;
};

/**
 * Execute Pivot Query Code Params
 *
 * @internal
 */
export type ExecutePivotQueryCodeParams = BaseCodeConfig & {
  pivotQueryParams: ExecutePivotQueryParams;
};

/**
 * Execute Query Code Props
 *
 * @internal
 */
export type ExecuteQueryCodeProps = Stringify<ExecuteQueryParams> & { extraImportsString: string };

/**
 * @internal
 */
export type ExecutePivotQueryCodeProps = Stringify<ExecutePivotQueryParams> & {
  extraImportsString: string;
};

/**
 * Chart Widget Code Props
 *
 * @internal
 */
export type ChartWidgetCodeProps = Stringify<ChartWidgetProps> & ExtraWidgetCodeProps;

/**
 * Pivot Table Widget Code Props
 *
 * @internal
 */
export type PivotTableWidgetCodeProps = Stringify<PivotTableWidgetProps> & ExtraWidgetCodeProps;

/**
 * Template Key by Widget Type
 *
 * @internal
 */
export type TemplateKeyMapByWidgetType = {
  chart: CodeTemplateKey;
  pivot: CodeTemplateKey;
  text?: CodeTemplateKey;
  custom?: CodeTemplateKey;
};

/** TYPE FOR DASHBOARD CODE */

/**
 * Dashboard Code Config
 *
 * @internal
 */
export type DashboardCodeConfig = BaseCodeConfig;

/**
 * Client-side Dashboard Code Params
 *
 * @internal
 */
export type ClientSideDashboardCodeParams = BaseCodeConfig & {
  dashboardProps: DashboardProps;
};

/**
 * By ID Dashboard Code Params
 *
 * @internal
 */
export type ByIdDashboardCodeParams = DashboardCodeConfig & {
  dashboardOid: string;
};

/**
 * Dashboard Code Params
 *
 * @internal
 */
export type DashboardCodeParams = ClientSideDashboardCodeParams | ByIdDashboardCodeParams;

/**
 * Check if dashboard code params is for client-side code
 *
 * @internal
 */
export const isClientSideDashboardCodeParams = (
  params: DashboardCodeParams,
): params is ClientSideDashboardCodeParams => {
  return 'dashboardProps' in params;
};

/**
 * Check if dashboard code params is for by ID code
 *
 * @internal
 */
export const isByIdDashboardCodeParams = (
  params: DashboardCodeParams,
): params is ByIdDashboardCodeParams => {
  return 'dashboardOid' in params;
};

/**
 * Dashboard Code Props
 *
 * @internal
 */
export type DashboardCodeProps = Stringify<DashboardProps> & ExtraWidgetCodeProps;

/**
 * @internal
 */
export type NormalizedField = {
  name: string;
  dataType: string;
  expression: string;
  description?: string;
};

/**
 * @internal
 */
export type NormalizedTable = {
  name: string;
  columns: NormalizedField[];
};

/**
 * Response JSON generated by the new NLQ API.
 *
 * @internal
 */
export type NlqResponseJSON = {
  dimensions: JSONArray;
  measures: JSONArray;
  filters: JSONArray;
  highlights?: JSONArray;
};

/**
 * Represents the result of a translation operation that can either succeed with data or fail with errors.
 *
 * @internal
 */
export type NlqTranslationResult<T> =
  | { success: true; data: T }
  | { success: false; errors: string[] };
