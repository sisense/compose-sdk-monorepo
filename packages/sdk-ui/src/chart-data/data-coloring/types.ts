/**
 * All possible color options for data.
 *
 * @see {@link https://docs.sisense.com/main/SisenseLinux/selecting-colors-in-widgets.htm | Selecting Colors in Widgets}
 */
export type DataColorOptions =
  | string
  | UniformDataColorOptions
  | RangeDataColorOptions
  | ConditionalDataColorOptions;

/**
 * Uniform color options for data similar to the Single Color option in the Sisense UI.
 *
 * @example
 * An example of specifying red as a uniform color for all data values.
 *
 * ```ts
 * {
 *   type: 'uniform',
 *   color: 'red',
 * }
 * ```
 * @see {@link https://docs.sisense.com/main/SisenseLinux/selecting-colors-in-widgets.htm | Selecting Colors in Widgets}
 */
export type UniformDataColorOptions = {
  /**
   * Color options type
   */
  type: 'uniform';
  /**
   * Color name, e.g., `red`, or a hexadecimal value, e.g., `#ff0000`.
   */
  color: string;
};

/**
 * Range color options for data similar to the Range option in the Sisense UI.
 *
 * Use `minColor` and `maxColor` to define the start and end color of the range.
 * A color name (for example, `red`), or a hexadecimal value (for example, `#ff0000`) can be specified.
 *
 * By default, the color range is set to match the minimum and maximum values of the data.
 * You can also override the default value range by
 * setting the `minValue`, `midValue`, and `maxValue` properties.
 *
 * @example
 * An example of specifying colors for data values ranging
 * from red (for min value) to blue (for max value)
 * with 2 more colors in between for a total of 4 colors:
 *
 * ```ts
 * {
 *   type: 'range',
 *   steps: 4,
 *   minColor: 'red',
 *   maxColor: 'blue',
 * }
 * ```
 * @see {@link https://docs.sisense.com/main/SisenseLinux/selecting-colors-in-widgets.htm | Selecting Colors in Widgets}
 */
export type RangeDataColorOptions = {
  /**
   * Color options type
   */
  type: 'range';
  /**
   * Distinct number of colors in the range
   */
  steps?: number;
  /**
   * Start color of the range
   */
  minColor?: string;
  /**
   * End color of the range
   */
  maxColor?: string;
  /**
   * Minimum value explicitly set to override the minimum value of the data
   */
  minValue?: number;
  /**
   * Middle value explicitly set to override the middle value of the data
   */
  midValue?: number;
  /**
   * Maximum value explicitly set to override the maximum value of the data
   */
  maxValue?: number;
};

/**
 * Conditional color options for data similar to the Conditional Color option in the Sisense UI.
 *
 * This option allows you to define color conditions.
 * Each condition is a logical expression that defines how data values are mapped into colors.
 * These conditions are evaluated in the order in which they appear in the array.
 *
 * @example
 * An example of a condition stating that a negative data value is displayed in red
 * and another condition stating that a positive data value is green.
 *
 * ```ts
 * {
 *   type: 'conditional',
 *   conditions: [
 *     { color: 'red', expression: '0', operator: '<' },
 *     { color: 'green', expression: '0', operator: '>=' },
 *   ],
 *   defaultColor: 'red',
 * }
 * ```
 * @see {@link https://docs.sisense.com/main/SisenseLinux/selecting-colors-in-widgets.htm | Selecting Colors in Widgets}
 */
export type ConditionalDataColorOptions = {
  /**
   * Color options type
   */
  type: 'conditional';
  /**
   * Array of color conditions
   */
  conditions?: DataColorCondition[];
  /**
   * Default color when no condition is met
   */
  defaultColor?: string;
};

/**
 * Color condition for {@link ConditionalDataColorOptions} represented as a logical expression.
 *
 * See {@link ConditionalDataColorOptions} for examples.
 *
 * @see {@link https://docs.sisense.com/main/SisenseLinux/selecting-colors-in-widgets.htm | Selecting Colors in Widgets}
 */
export type DataColorCondition = {
  /**
   * Color for this condition
   */
  color: string;
  /**
   * Expression representing the data value
   */
  expression: string;
  /**
   * Supported operators for `expression`
   */
  operator: '<' | '>' | '≤' | '<=' | '≥' | '>=' | '=' | '≠' | '!=';
};
