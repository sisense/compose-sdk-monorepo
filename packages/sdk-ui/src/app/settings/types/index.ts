export type { GlobalsObject } from './globals.js';
