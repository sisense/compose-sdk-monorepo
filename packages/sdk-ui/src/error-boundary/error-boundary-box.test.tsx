import ErrorBoundaryBox from './error-boundary-box';
import { fireEvent, render } from '@testing-library/react';
import { ThemeContext } from '@/theme-provider';
import { CompleteThemeSettings } from '@/types';

describe('ErrorBoundaryBox', () => {
  it('should render error correctly', () => {
    const { container } = render(<ErrorBoundaryBox error="error" />);
    expect(container).toMatchSnapshot();
  });

  it('should render error with theme settings correctly', () => {
    const { container } = render(
      <ThemeContext.Provider
        value={{
          themeSettings: {
            general: {
              backgroundColor: 'red',
              brandColor: 'blue',
            },
            typography: {
              fontFamily: 'Roboto',
              primaryTextColor: 'green',
            },
          } as CompleteThemeSettings,
        }}
      >
        <ErrorBoundaryBox error="error" />
      </ThemeContext.Provider>,
    );

    fireEvent.mouseOver(container.firstChild as Element);

    expect(container).toMatchSnapshot();
  });
});
