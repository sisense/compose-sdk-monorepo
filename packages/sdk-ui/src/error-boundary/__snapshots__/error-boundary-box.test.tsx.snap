// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundaryBox > should render error correctly 1`] = `
<div>
  <div
    aria-label="error-box"
    class="_container_5eeb40"
  >
    <div
      class="_card_5eeb40"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="csdk-flex csdk-justify-center csdk-mb-[10px]"
      >
        <svg
          height="53px"
          viewBox="0 0 53 53"
          width="53px"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g>
              <polygon
                fill="#ffcb05"
                points="26.4996 -0.000100000005 -0.000400000001 26.5009 26.4996 52.9999 53.0006 26.5009"
              />
              <path
                d="M24.5,39.054 L24.5,35.054 L28.5,35.054 L28.5,39.054 L24.5,39.054 Z M28.5,31.0536 L24.5,31.0536 L23.5,13.0536 L29.5,13.0536 L28.5,31.0536 Z"
                fill="#FFFFFF"
              />
            </g>
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundaryBox > should render error with theme settings correctly 1`] = `
<div>
  <div
    aria-label="error-box"
    class="_container_5eeb40"
  >
    <div
      class="_card_5eeb40"
      style="background-color: red;"
    >
      <div
        class="csdk-flex csdk-justify-center csdk-mb-[10px]"
      >
        <svg
          height="53px"
          viewBox="0 0 53 53"
          width="53px"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g>
              <polygon
                fill="blue"
                points="26.4996 -0.000100000005 -0.000400000001 26.5009 26.4996 52.9999 53.0006 26.5009"
              />
              <path
                d="M24.5,39.054 L24.5,35.054 L28.5,35.054 L28.5,39.054 L24.5,39.054 Z M28.5,31.0536 L24.5,31.0536 L23.5,13.0536 L29.5,13.0536 L28.5,31.0536 Z"
                fill="#FFFFFF"
              />
            </g>
          </g>
        </svg>
      </div>
      <div
        class="_text_5eeb40"
        style="color: green; font-family: Roboto;"
      >
        errorBoxText
      </div>
    </div>
  </div>
</div>
`;
